vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd

```diff
@@ -0,0 +1,969 @@
+
+One of the central abstraction in Keras is the `Layer` class. A layer
+encapsulates both a state (the layer's "weights") and a transformation from
+inputs to outputs (a "call", the layer's forward pass).
+
+Here's a densely-connected layer. It has a state: the variables `w` and `b`.
+```{r}
+Linear(keras$layers$Layer) %py_class% {
+  initialize <- function(units = 32, input_dim = 32) {
+    super$initialize()
+    w_init <- tf$random_normal_initializer()
+    self$w <- tf$Variable(
+      initial_value = w_init(
+        shape = shape(input_dim, units),
+        dtype = "float32"
+      ),
+      trainable = TRUE
+    )
+    b_init <- tf$zeros_initializer()
+    self$b <- tf$Variable(
+      initial_value = b_init(shape = shape(units), dtype = "float32"),
+      trainable = TRUE
+    )
+  }
+
+  call <- function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+}
+```
+
+
+You would use a layer by calling it on some tensor input(s), much like a regular
+function.
+
+```{r}
+x <- tf$ones(shape(2, 2))
+linear_layer <- Linear(4, 2)
+y <- linear_layer(x)
+print(y)
+```
+
+All the builtin layers in the keras package (e.g., `layer_dense()`) are
+additionally designed to compose nicely with the pipe operator (`%>%`), so that
+the layer instance is conveniently created on demand when an existing model or
+tensor is piped in. In order to make a custom layer similarly compose nicely
+with the pipe, you can call `create_layer_wrapper()`.
+
+```{r}
+layer_linear <- create_layer_wrapper(Linear)
+layer_linear
+# function (object, units = 32, input_dim = 32)
+# {
+#     args <- capture_args(match.call(), modifiers)
+#     create_layer(LayerClass, object, args)
+# }
+# <environment: 0x557cc3d71158>
+```
+Now `layer_linear` is a layer constructor composes nicely with `%>%`, just like the built in layers:
+```{r}
+model <- keras_model_sequential() %>%
+  layer_linear(4, 2)
+model(k_ones(c(2, 2))) |> invisible() # force model to build
+model
+# Model
+# Model: "sequential"
+# __________________________________________________________________________________________
+# Layer (type)                            Output Shape                        Param #
+# ==========================================================================================
+# linear_1 (Linear)                       (2, 4)                              12
+# ==========================================================================================
+# Total params: 12
+# Trainable params: 12
+# Non-trainable params: 0
+# __________________________________________________________________________________________
+```
+
+
+Note that the weights `w` and `b` are automatically tracked by the layer upon
+being set as layer attributes:
+
+```{r}
+stopifnot(all.equal(
+  linear_layer$weights,
+  list(linear_layer$w, linear_layer$b)
+))
+```
+
+Note you also have access to a quicker shortcut for adding weight to a layer:
+the `add_weight()` method:
```

Comment:
i think that if the guide wasn't a translation, i'd only show this way of doing it ...